SecurityPkg/VariableAuthenticated/SecureBootConfigDxe/SecureBootConfigImpl.c

```diff
@@ -4602,6 +4733,51 @@ KeyEraseAll (
   return Status;
 }
 
+/**
+
+  Check whether a reset is needed, if reset is needed, Popup a menu to notice user.
+
+**/
+VOID
+SetupResetReminder (
+  VOID
+  )
+{
+  EFI_INPUT_KEY                           Key;
+  CHAR16                                  *StringBuffer1;
+  CHAR16                                  *StringBuffer2;
+  EFI_STATUS                              Status;
+  EDKII_FORM_BROWSER_EXTENSION2_PROTOCOL  *FormBrowserEx2;
+
+  //
+  // Use BrowserEx2 protocol to check whether reset is required.
+  //
+  Status = gBS->LocateProtocol (&gEdkiiFormBrowserEx2ProtocolGuid, NULL, (VOID **)&FormBrowserEx2);
+
+  //
+  // check any reset required change is applied? if yes, reset system
+  //
+  if (!EFI_ERROR (Status) && FormBrowserEx2->IsResetRequired ()) {
+    StringBuffer1 = AllocateZeroPool (MAX_CHAR * sizeof (CHAR16));
+    ASSERT (StringBuffer1 != NULL);
+    StringBuffer2 = AllocateZeroPool (MAX_CHAR * sizeof (CHAR16));
+    ASSERT (StringBuffer2 != NULL);
+    StrCpyS (StringBuffer1, MAX_CHAR, L"Configuration changed. Reset to apply it Now.");
+    StrCpyS (StringBuffer2, MAX_CHAR, L"Press ENTER to reset");

```

Comment:
Why allocate memory instead of just passing literals?

Reply:
I guess it is just replicated inefficiency from:

- MdeModulePkg/Application/UiApp/FrontPage.c (SetupResetReminder)
- MdeModulePkg/Library/BootMaintenanceManagerUiLib/BootOption.c (BmmSetupResetReminder)
- MdeModulePkg/Library/BootManagerUiLib/BootManager.c (BmSetupResetReminder)

I can change them all basically.

Reply:
8ec13e236160fa03941868464afd7ae0f8134bb6